projects/pgai/pgai/semantic_catalog/gen_sql.py

```diff
@@ -554,127 +719,166 @@ async def generate_sql(
-                    ctx=ctx,
-                    prompt=prompt,
-                    prior_prompts=prior_prompts,
-                    error=error,
-                )
-            else:
-                break  # we have a valid answer
-        else:
-            user_prompt = None
+        messages.append(model_response)
+        usage = usage + model_response.usage
+
+        for part in model_response.parts:
+            match part.part_kind:
+                case "tool-call":
+                    tool_call_part: ToolCallPart = part
+                    args = tool_call_part.args_as_dict()
+                    logger.info(f"tool call: {tool_call_part.tool_name}")
+                    if tool_call_part.tool_name == _SEARCH_TOOL_NAME:
+                        questions: list[str] = (
+                            cast(list[str], args.get("questions")) or []
+                        )
+                        for q in questions:
+                            prior_prompts.append(q)
+                            logger.info(f"semantic search for '{q}'")
+                            ctx = await fetch_database_context(
+                                catalog_con,
+                                target_con,
+                                catalog_id,
+                                embedding_name,
+                                embedding_config,
+                                q,
+                                ctx,
+                                sample_size=sample_size,
+                            )
+                    elif tool_call_part.tool_name == _RECORD_TOOL_NAME:
+                        answer = args.get("sql_statement")
+                        if not answer:
+                            logger.error("llm did not provide a sql_statement")
+                            continue
+                        _continue = False
+                        relevant_object_ids: list[int] | None = cast(
+                            list[int] | None, args.get("relevant_object_ids")
+                        )
+                        relevant_sql_example_ids: list[int] | None = cast(
+                            list[int] | None, args.get("relevant_sql_example_ids")
+                        )
+                        relevant_fact_ids: list[int] | None = cast(
+                            list[int] | None, args.get("relevant_fact_ids")
+                        )
+                        if relevant_object_ids:
+                            for irrelevant_id in (
+                                ctx.objects.keys() - relevant_object_ids
+                            ):
+                                ctx.objects.pop(irrelevant_id)
+                                ctx.rendered_objects.pop(irrelevant_id)
+                            await _increment_obj_usage(
+                                catalog_con,
+                                catalog_id,
+                                [id for id in ctx.objects],
+                            )
```

Comment:
Would it make sense to wait to do this block after we'e confirmed that there's no syntax errors in the query?

Minimally I think we should wait to do any increment till after we've verified that.